pkg/apis/messaging/v1alpha1/in_memory_channel_lifecycle.go

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2019 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+var imcCondSet = duckv1alpha1.NewLivingConditionSet(InMemoryChannelConditionDispatcherReady, InMemoryChannelConditionServiceReady, InMemoryChannelConditionEndpointsReady, InMemoryChannelConditionAddressable)
+
+const (
+	// InMemoryChannelConditionReady has status True when all subconditions below have been set to True.
+	InMemoryChannelConditionReady = duckv1alpha1.ConditionReady
+
+	// InMemoryChannelConditionDispatcherReady has status True when a Dispatcher deployment is ready
+	// Keyed off appsv1.DeploymentAvaialble, which means minimum available replicas required are up
+	// and running for at least minReadySeconds.
+	InMemoryChannelConditionDispatcherReady duckv1alpha1.ConditionType = "DispatcherReady"
+
+	// InMemoryChannelConditionServiceReady has status True when a k8s Service is ready. This
+	// basically just means it exists because there's no meaningful status in Service. See Endpoints
+	// below.
+	InMemoryChannelConditionServiceReady duckv1alpha1.ConditionType = "ServiceReady"
+
+	// InMemoryChannelConditionEndpointsReady has status True when a k8s Service Endpoints are backed
+	// by at least one endpoint.
+	InMemoryChannelConditionEndpointsReady duckv1alpha1.ConditionType = "EndpointsReady"
+
+	// InMemoryChannelConditionAddressable has status true when this InMemoryChannel meets
+	// the Addressable contract and has a non-empty hostname.
+	InMemoryChannelConditionAddressable duckv1alpha1.ConditionType = "Addressable"
+)
+
+// GetCondition returns the condition currently associated with the given type, or nil.
+func (imcs *InMemoryChannelStatus) GetCondition(t duckv1alpha1.ConditionType) *duckv1alpha1.Condition {
+	return imcCondSet.Manage(imcs).GetCondition(t)
+}
+
+// IsReady returns true if the resource is ready overall.
+func (imcs *InMemoryChannelStatus) IsReady() bool {
+	return imcCondSet.Manage(imcs).IsHappy()
+}
+
+// InitializeConditions sets relevant unset conditions to Unknown state.
+func (imcs *InMemoryChannelStatus) InitializeConditions() {
+	imcCondSet.Manage(imcs).InitializeConditions()
+}
+
+func (imcs *InMemoryChannelStatus) SetAddress(hostname string) {
+	imcs.Address.Hostname = hostname
```

Comment:
We could switch to use URI (the new addressable field that is added in beta apis) rather than hostname. This could be done separately as a different issue. Just mentioned it here in case you think it is easy to change it as part of this bigger CRD change, then can accommodate it here.

Reply:
Added a TODO. I'll change this once I start playing with the other duck types too.